src/cpu/core_dyn_x86.cpp

```diff
@@ -263,6 +263,31 @@ Bits CPU_Core_Dyn_X86_Run(void) {
 	};
 	auto_dh_fpu fpu_saver;
 
+	class auto_fpu_sync {
+	public:
+		auto_fpu_sync () {
+			FPU_SetTag(dyn_dh_fpu.state.tag);
```

Comment:
@finalpatch , I wonder if there is/was already existing code to do something (almost?) like this already?

![2023-01-24_15-54](https://user-images.githubusercontent.com/1557255/214460510-12bf9626-89d5-4ce4-9443-837079a8382e.png)

Reply:
These are different.  The `auto_dh_fpu` class saves the hardware fpu state to `dyn_dh_fpu.state`.
The code I added then synchronizes `dyn_dh_fpu.state` with `fpu.*`

Reply:
there are 3 sets of fpu states when the dyn_x86 core with dh_fpu is running:
1. The physical fpu that runs the recompiled host code
2. `dh_fpu`,  this stores a snapshot of the physical fpu whenever the core leaves recompiled host code, so that it can resume later.
3. `fpu`, this is the state that the normal core, and non-dh fpu codegen uses.

My patch syncs 2 and 3 so that dyn core and normal core can interleave with each other, which they do

Reply:
Thanks for explaining this, @finalpatch. 

I think the name could be made clearer (and it's now very tangled up with the existing FPU saver):

```c++
auto_dh_fpu // existing
auto_fpu_sync // new
```

I think the functionality could be made more explicit with two static functions instead of lazier "we restore whenever the object goes out of scope".

```c++
write_normal_fpu_from_dyn(...);
CPU_Core_Normal_Run();
read_normal_fpu_to_dyn(...);
```

Where both those become local static functions, without the class wrapper.

Reply:
I would rather not touch the existing code because like I said, they do different things, and they work.  I can comment this part so that later readers understand why we need these things.

Reply:
> I would rather not touch the existing code because like I said, they do different things, and they work. I can comment this part so that later readers understand why we need these things.

I agree; I'm just commenting on the added class, in this PR `auto_fpu_sync`.  I think it would be better implemented not as a class, but as two static functions: a loader and a saver, that get called directly before and after the normal core is run.

the existing "FPU saver" can stay as-is.

Reply:
For these kind of things that must be used in pairs,  I think writing them as class and relying on the compiler calling ctor/dtor is better.  If later the `CPU_Core_Normal_Run()` function decides to throw exceptions, we are still good with the ctor/dtor style.

Reply:
>  I think writing them as class and relying on the compiler calling ctor/dtor is better. If later the CPU_Core_Normal_Run() function decides to throw exceptions, we are still good with the ctor/dtor style.

That's a good point - especially when the intent is to clean up at some future time.  For example, I can use a std::vector locally or as member variables, and eventually they're mopped up whenever they're scope closes.

In this case, we're not just mopping up memory.  Future developers aren't free to place this object in a member variable our higher up in the call loop.  Doing so would mean the scope lasts for several rounds of the normal core.

We _really_ do want the dyn FPU restored right after the normal core comes back. I've got two thoughts:

explicit scoping, like the scoped c++17 mutexes:

``` c++
} else {
	int32_t old_cycles=CPU_Cycles;
	CPU_Cycles=1;
	Bits nc_retcode = 0;

	{  // explicit scope for FPU state restore-on-d'tor
		fpu_saver = auto_dh_fpu();
		auto_fpu_sync sync_fpu; 
		nc_retcode = CPU_Core_Normal_Run();
	}
	
	...
```

Reply:
The other thought is to have an explicit restore member function that we call explicitly:

```c++
auto_fpu_sync sync_fpu; 
nc_retcode = CPU_Core_Normal_Run();
sync_fpu.restore();
```

Which also gets conditionally called on destruction, if it hasn't been called yet.

this approach is also similar to the scoped mutex (which also has a safety net if it goes out of scope).

Reply:
I moved the fpu sync into a local scope.